src/Vertex.java

```diff
@@ -1,15 +1,45 @@
+import java.util.Objects;
+
 public class Vertex {
 
-    public int x;
-    public int y;
+    private final int x;
+    private final int y;
 
     public Vertex(int x, int y) {
         this.x = x;
         this.y = y;
     }
 
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) return true;
+
+        if (other == null || getClass() != other.getClass()) return false;
+
+        Vertex vertex = (Vertex) other;
+
+        return x == vertex.x && y == vertex.y;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(x, y);
+    }
+
     @Override
     public String toString() {
         return "[" + x + "," + y + "]";
     }
+
+    public int getX() {
+        return x;
+    }
+
+    public int getY() {
+        return y;
+    }
+
+    public boolean isAdjacent(Vertex other) {
+        return (x == other.x + 1 && y == other.y) || (x == other.x && y == other.y + 1);
```

Comment:
The issue was caused by us not checking that the y co-ordinates are the same when the x co-ordinates are adjacent or that the x co-ordinates are the same when the y co-ordinates are adjacent. This means that it would return [0,1], [1,1], [2,1], [3,1], [4,1] etc. as being adjacent to [0,0], instead of only [0,1] and [1,0] as expected.